subt_ign/src/path_tracer.cc

```diff
@@ -0,0 +1,354 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#include "path_tracer.hh"
+
+//////////////////////////////////////////////////
+Processor::Processor(const std::string &_path, int _stepSleepMs)
+  : stepSleepMs(_stepSleepMs)
+{
+  // Create the transport node with the partition used by simulation
+  // world.
+  ignition::transport::NodeOptions opts;
+  opts.SetPartition("PATH_TRACER");
+  this->markerNode = new ignition::transport::Node(opts);
+  this->ClearMarkers();
+
+  // Subscribe to the artifact poses.
+  this->SubscribeToArtifactPoseTopics();
+
+  // Playback the log file.
+  std::unique_lock<std::mutex> lock(this->mutex);
+  std::thread playbackThread(std::bind(&Processor::Playback, this, _path));
+
+  this->cv.wait(lock);
+
+  // Create a transport node that uses the default partition.
+  ignition::transport::Node node;
+
+  // Subscribe to the robot pose topic
+  bool subscribed = false;
+  for (int i = 0; i < 5 && !subscribed; ++i)
+  {
+    std::vector<std::string> topics;
+    node.TopicList(topics);
+
+    // Subscribe to the first /dynamic_pose/info topic
+    for (auto const &topic : topics)
+    {
+      if (topic.find("/dynamic_pose/info") != std::string::npos)
+      {
+        // Subscribe to a topic by registering a callback.
+        if (!node.Subscribe(topic, &Processor::Cb, this))
+        {
+          std::cerr << "Error subscribing to topic ["
+            << topic << "]" << std::endl;
+          return;
+        }
+        subscribed = true;
+        break;
+      }
+    }
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+  }
+
+  // Wait for log playback to end.
+  playbackThread.join();
+
+  // Process the events log file.
+  YAML::Node events = YAML::LoadFile(_path + "/events.yml");
+  for (std::size_t i = 0; i < events.size(); ++i)
+  {
+    if (events[i]["type"].as<std::string>() == "artifact_report_attempt")
+    {
+      ignition::math::Vector3d reportedPos;
+
+      // Read the reported pose.
+      std::stringstream stream;
+      stream << events[i]["reported_pose"].as<std::string>();
+      stream >> reportedPos;
+
+      int sec = events[i]["time_sec"].as<int>();
+      std::unique_ptr<ReportData> data(new ReportData);
```

Comment:
Prefer `make_unique` vs `new`.

```suggestion
      auto data = std::make_unique<ReportData>();
```

Reply:
Fixed in 4890364ef47509e855e24a38e3fd5e5b3a6da1c5